tests/commands/test_get.py

```diff
@@ -0,0 +1,432 @@
+    output = [output.split('\t') for output in ret.stdout.split('\n')][:-1]
+
+    for index, key in enumerate(keys or ['type']):
+        key, index = (key, index) if sort in ['-s', '-S'] else ('path', -1)
+        sorted_values = presorted.get(key, ['<unset>'] * 3)
+        print(key)
+        values = [line[index] for line in output]
+
+        if sort == '-S':
+            sorted_values = list(reversed(sorted_values))
+
+        assert values == sorted_values
+
+    assert not ret.stderr
+    assert ret.returncode == 0
+
+
+@pytest.mark.parametrize('keys', [None, ['num'], ['str', 'num']])
+@pytest.mark.parametrize('reverse', [True, False])
+def test_natural_sort(
+        keys: Union[list, None], reverse: bool) -> None:
+    assets = [
+        (Path('a13bc_foo_bar.1'), {'num': 'num-20', 'str': 'abc', 'id': '1'}),
+        (Path('a2cd_foo_bar.2'), {'num': 'num-3', 'str': 'def', 'id': '2'}),
+        (Path('a36ab_foo_bar.3'), {'num': 'num-16', 'str': 'ghi', 'id': '3'})]
+    sorted_assets = natural_sort(assets, keys=keys, reverse=reverse)
+    ids = [data.get('id') for _, data in sorted_assets]
+
+    if reverse:
+        ids = list(reversed(ids))
+
+    if keys is None:
+        assert ids == ['2', '1', '3']
+    elif keys[0] == 'num':
+        assert ids == ['2', '3', '1']
+    elif keys[0] == 'str':
+        assert ids == ['1', '2', '3']
+
+
+def test_fill_unset() -> None:
+    assets = [
+        (Path('a13bc_foo_bar.1'), {'num': 'num-20', 'str': 'abc'}),
+        (Path('a2cd_foo_bar.2'), {'num': 'num-3'}),
+        (Path('a36ab_foo_bar.3'), {'str': 'ghi'})]
+    keys = ['type', 'make', 'model', 'serial', 'num', 'str', 'id']
+    filled = list(fill_unset((a for a in assets), keys=keys))
+
+    for i, (asset, data) in enumerate(filled):
+        assert isinstance(asset, Path)
+        assert asset == assets[i][0]
+        for k, v in data.items():
+            assert v == assets[i][1].get(k, '<unset>')
+
+    assert filled[1][1]['str'] == '<unset>'
+    assert filled[2][1]['num'] == '<unset>'
+
+
+@pytest.mark.parametrize('depth', [None, '2', 2])
+def test_eval_depth(depth: Union[str, int, None]) -> None:
+    assert eval_depth(depth, default=2) == 2
+
+
+@pytest.mark.parametrize('filters', [
+    ['type=laptop'], ['type=laptop', 'make=foo', 'bar=1']])
+def test_eval_filters(filters: list[str]) -> None:
+    class Repo:
+        pass
+
+    validated_filters = eval_filters(filters, repo=Repo())  # pyre-ignore[6]
+    filter_dict = {k: v for k, v in [f.split('=', 1) for f in filters]}
+
+    for f in validated_filters:
+        assert f.value == filter_dict[f.key]
+
+    assert len(validated_filters) == len(filters)
+    assert all(isinstance(f, Filter) for f in validated_filters)
+
+
+@pytest.mark.parametrize('filters', [
+    ['type=laptop', 'type=laptop'], ['badfilter']])
+@pytest.mark.parametrize('rich', [True, False])
+def test_eval_filters_error(capsys, filters: list[str], rich: bool) -> None:
```

Comment:
Did not know that, thanks for the heads up! I'll add it in right away.

Reply:
I do not see why you can't do what I describe and then pass the repo-object into the filter like any other variable? The repo from ` def test_eval_filters_error(repo: Repo, capsys, filters: list[str], rich: bool) -> None:` is exactly what your answer mentions that is needed, e.g. "the current `Repo` instance".

Reply:
`repo: Repo` is not a repo instance, but a reference to the `Repo` class. There's a big difference between them. Type hinting won't complain because a `Repo` instance is of the same type as the `Repo` class. In this particular case it will work, because part of `Filter` I'm testing never touches `Filter.repo`, but in my opinion it is an uncomfortable precedence to set. Since it works for filter and you seem set on avoiding mocking here, I will add this, but I advise caution using this approach elsewhere.

```
class Test:
    def __init__(self, foo: str) -> None:
        self.foo = foo

foo = Test('bar')
```

We can see the difference as:
```
>>> foo
<__main__.Test object at xxx>
>>> Repo
<class '__main__.Repo'>
```

They behave differently.